File: mylibrary/src/main/java/com/xiaolu/mylibrary/base/BaseMvpFragment.java
package com.xiaolu.mylibrary.base;

import android.os.Bundle;

import androidx.viewbinding.ViewBinding;

import com.xiaolu.mylibrary.mvpbase.BaseContract;
import com.xiaolu.mylibrary.mvpbase.BasePresenter;
import com.xiaolu.mylibrary.mvpbase.IPresenter;
import com.xiaolu.mylibrary.mvpbase.IView;
import com.xiaolu.mylibrary.mvpbase.MvpCallback;
import com.xiaolu.mylibrary.net.RxManager;


/**
 * @author: zhaol
 * @createdOn: 2018/5/17 16:19
 * @description: $desc$
 */
public abstract class BaseMvpFragment<VB extends ViewBinding, V extends IView, P extends IPresenter<V>> extends RXBaseFragment<VB>
        implements MvpCallback<V, P> {
    protected P mPresenter;
    protected V mView;

    @Override
    public void onActivityCreated(Bundle savedInstanceState) {
        super.onActivityCreated(savedInstanceState);
        mView = createView();
        if (getPresenter() == null) {
            mPresenter = createPresenter();
            try {
                mPresenter.attachView((V) this);
                initDate();
            } catch (Exception e) {
                new ClassCastException(this.toString() + "实现BasePresenter或者BaseView子类接口");
            }
        }
    }

    @Override
    public P getPresenter() {
        return mPresenter;
    }

    @Override
    public void setPresenter(P presenter) {
        this.mPresenter = presenter;
    }

    @Override
    public V getMvpView() {
        return mView;
    }

    @Override
    public void setMvpView(V view) {
        this.mView = view;
    }

    @Override
    public void onDestroy() {
        if (mPresenter != null) {
            RxManager.getInstance().clear(mPresenter.getClass().getName());
        }
        if (mPresenter != null) {
            setMvpView(null);
            setPresenter(null);
        }
        super.onDestroy();
    }

    @Override
    public void onDestroyView() {
        if (mPresenter != null) {
            setMvpView(null);
            setPresenter(null);
        }
        super.onDestroyView();
    }

    @Override
    public void onResume() {
        if (mPresenter != null) {
            mPresenter.attachView((V) this);
        } else {
            mPresenter = createPresenter();
        }
        super.onResume();
    }
}

File: mylibraryKotlin/src/main/java/com/xiaolu/mylibrarykotlin/base/BaseMvpFragment.kt
package com.xiaolu.mylibrarykotlin.base

import android.os.Bundle
import com.xiaolu.mylibrarykotlin.mvpbase.IPresenter
import com.xiaolu.mylibrarykotlin.mvpbase.IView
import com.xiaolu.mylibrarykotlin.mvpbase.MvpCallback
import com.xiaolu.mylibrarykotlin.net.RxManager

/**
 * @author: zhaol
 * @createdOn: 2018/5/17 16:19
 * @description: $desc$
 */
abstract class BaseMvpFragment<V : IView?, P : IPresenter<V?>?> : RXBaseFragment(), MvpCallback<V?, P?> {
    override var presenter: P? = null
    override var mvpView: V? = null
    override fun onActivityCreated(savedInstanceState: Bundle?) {
        super.onActivityCreated(savedInstanceState)
        mvpView = createView()
        if (presenter == null) {
            presenter = createPresenter()
            try {
                presenter!!.attachView(this as V)
                initDate()
            } catch (e: Exception) {
                ClassCastException(this.toString() + "实现BasePresenter或者BaseView子类接口")
            }
        }
    }

    override fun onDestroy() {
        RxManager.instance?.clear(BaseMvpActivity::class.simpleName)
        if (presenter != null) {
            mvpView = null
            presenter = null
        }
        super.onDestroy()
    }

    override fun onDestroyView() {
        if (presenter != null) {
            mvpView = null
            presenter = null
        }
        super.onDestroyView()
    }

    override fun onResume() {
        if (presenter != null) {
            presenter?.attachView(this as V)
        } else {
            presenter = createPresenter()
        }
        super.onResume()
    }
}